File: lib/py/lib_bejson_gaming/lib_bejson_assets.py
"""
game1 aka the BEJSON Game Engine (#1)
Author: Elton Boehnen
"""
import json

class BEJSONAssetRegistry:
    """
    Mirror of lib_bejson_assets.js
    Python-compatible BEJSON-Native Asset Registry (104a).
    """
    def __init__(self, name="AssetRegistry"):
        self.bejson = {
            "Format": "BEJSON",
            "Format_Version": "104a",
            "Schema_Name": name,
            "Records_Type": ["Asset"],
            "Fields": [
                { "name": "id", "type": "string" },
                { "name": "type", "type": "string" },
                { "name": "path", "type": "string" },
                { "name": "loaded", "type": "boolean" }
            ],
            "Values": []
        }
        self.cache = {}

    def register_asset(self, asset_id, asset_type, path):
        """Adds or updates an asset in the registry."""
        idx = next((i for i, v in enumerate(self.bejson["Values"]) if v[0] == asset_id), -1)
        if idx != -1:
            self.bejson["Values"][idx] = [asset_id, asset_type, path, False]
        else:
            self.bejson["Values"].append([asset_id, asset_type, path, False])

    def mark_loaded(self, asset_id, loaded=True):
        idx = next((i for i, v in enumerate(self.bejson["Values"]) if v[0] == asset_id), -1)
        if idx != -1:
            self.bejson["Values"][idx][3] = loaded

    def export_bejson(self):
        return json.dumps(self.bejson, indent=2)

File: lib/js/lib_bejson_gaming/lib_bejson_assets.js
/**
 * switch_assets.js
 * game1 aka the BEJSON Game Engine (#1)
 * Author: Elton Boehnen: BEJSON-Native Asset Registry (v1.0)
 * Date: 2026-05-03
 */

window.Switch = window.Switch || {};

class SwitchAssets {
    constructor(name = "AssetRegistry") {
        this.bejson = Switch.BEJSON.create104a(name, [
            { name: "id", type: "string" },
            { name: "type", type: "string" },
            { name: "path", type: "string" },
            { name: "loaded", type: "boolean" }
        ], []);
        this.cache = new Map();
    }

    async load(id, type, path) {
        if (this.cache.has(id)) return this.cache.get(id);
        let asset;
        if (type === 'image') asset = await this._loadImage(path);
        else if (type === 'json') asset = await (await fetch(path)).json();
        this.cache.set(id, asset);
        return asset;
    }

    _loadImage(path) {
        return new Promise((res, rej) => {
            const img = new Image();
            img.onload = () => res(img);
            img.onerror = rej;
            img.src = path;
        });
    }

    get(id) { return this.cache.get(id); }
}

Switch.Assets = SwitchAssets;
